twine-upload.sh

```diff
@@ -17,6 +17,16 @@ export PATH="/usr/bin:${PATH}"  # To find `id`
 export PATH="$(python -m site --user-base)/bin:${PATH}"
 export PYTHONPATH="$(python -m site --user-site):${PYTHONPATH}"
 
+if [[ -z "${INPUT_PASSWORD}" ]] ; then
+    # No password supplied by the user implies that we're in the OIDC flow;
+    # retrieve the OIDC credential and exchange it for a PyPI API token.
+    echo "::notice::in OIDC flow"
+    INPUT_PASSWORD=$(python /app/oidc-exchange.py)
+
+    # Mask the newly minted PyPI token, so that we don't accidentally leak it
+    # in logs.
+    echo "::add-mask::${INPUT_PASSWORD}"
```

Comment:
You can print this from the Python module too, it'll probably be a bit safer this way. I'd do so immediately upon reading the API response to prevent it from appearing in traces accidentally.

Reply:
Yeah -- that's what I had originally, but I wasn't sure how it would interact with the subshell (i.e. the fact that we ultimately retrieve the token by doing `INPUT_PASSWORD=$(...)`).

I agree it's much better inside the Python script, I'll experiment.

Reply:
Ah, I haven't thought of that. Ultimately, I'd like to get rid of sh at some point but haven't prioritized that yet.
FWIW, these workflow commands work if printed to stderr — I checked. But yes, manual validation is a good idea either way.

Reply:
> FWIW, these workflow commands work if printed to stderr — I checked.

Nice! I'll do that.

>  Ultimately, I'd like to get rid of sh at some point but haven't prioritized that yet.

I'm happy to help with this, when the time comes :slightly_smiling_face: -- https://github.com/pypa/gh-action-pip-audit may be a good reference point, since we started out with 100% Python for it.

Reply:
Yes, I know how to do it, just haven't had a chance to start. I did that for `alls-green`, for example.
This one, though, has other assumptions. When I started, composite didn't exist, later it wasn't fully usable. But also, I wanted to explicitly manage the env and not influence/break whatever env the end-users may already have in their jobs. This is why I've gone for docker. I was also thinking about pre-caching the image so it'd be like 30 seconds faster.
Though, as a side effect, this helped promote the idea that publishing should be done from a single job and not spread out across many parallel ones. It annoys somebody but I think it is also letting them structure their CI/CD better. Switching to composite would mean opening up a can of worms in a way.
So the migration to Python would need to keep docker. At least, at the beginning. I really care about not breaking the surrounding env and promoting the best practices. Plus I feel like the caching would be better if implemented...